README.md

# ClusterMon

ocf_heartbeat_ClusterMon (7) - Runs crm_mon in the background, recording the cluster status to an HTML file.

A Pacemaker cluster is an event driven system. In this context, an event is a resource failure or configuration change (not exhaustive). Parameter and details can be found in the man page, under <http://linux-ha.org/doc/man-pages/re-ra-ClusterMon.html> or with crm:

    crm(live)ra# info ClusterMon

The `ocf:heartbeat:ClusterMon` resource can monitor the cluster status and triggers alerts on each cluster event. This resource runs crm_mon in the background at regular intervals (configurable) and uses crm_mon capabilities to send emails (SMTP), SNMP traps or to execute an external program via the extra_options parameter. It works by using crm_mon in the background, which is a binary that provides a summary of cluster’s current state. This binary has a couple options to send email (SMTP) or traps (SNMP) on any transition to a chosen recipient. You need a crm_mon binary that supports sending SNMP or SMTP!

## Helper Script

At least it seems that SLES15 has dropped the mail-to option from crm_mon. So we need to write a workaround around that using a little helper script :-(

### SLES12+SLES15 and a helper script 
(crm_mail_agent.sh)

    primitive rsc_ClusterMon ocf:pacemaker:ClusterMon \
      params user=root update=10000 pidfile="/crm_scripts/crm_monitor/crmMon.pid" \
      htmlfile="/crm_scripts/crm_monitor/crmHtml.html"  \
      extra_options="-E /crm_scripts/crm_monitor/crm_mail_agent.sh" \
      op monitor on-fail=restart interval=60
    clone ClusterMon-clone rsc_ClusterMon \
      meta target-role=Started

This package/file is hosted on <https://github.com/roseswe/ClusterMon>

### SLES11 and maybe other distros 
(Feedback highly appreciated)

    primitive rsc_ClusterMon ocf:pacemaker:ClusterMon \
        params user=root update=30 extra_options="--mail-to=root" \
        op monitor on-fail=restart interval=60
        clone ClusterMon-clone rsc_ClusterMon \
        meta target-role=Started


<!-- vim:set fileencoding=utf8 fileformat=unix filetype=gfm tabstop=2 expandtab: 
@(#)$Id: README.md,v 1.1 2022/03/02 14:37:41 ralph Exp $  -->


ChangeLog.md
Changelog
=========


(unreleased)
------------

Fix
~~~
- MarkDown errors fixed chg: added CVS keywords. [roseswe]


1.7.0 (2022-03-02)
------------------
- Add: Initial creaton of README.md. [Ralph Roth]




crm_mail_agent.sh
#!/bin/bash
# vim:set fileencoding=utf8 fileformat=unix filetype=shell tabstop=2 expandtab:
# @(#)$Id: crm_mail_agent.sh,v 1.7 2022/03/02 14:37:41 ralph Exp $
# SLES15 and SLES12 dropped the --mail-to option in crm_mon, so we have to emulate this with this script

# Example: /crm_scripts/crm_monitor/crm_mail_agent.sh
# this resolves to
# /usr/sbin/crm_mon -p /tmp/ClusterMon_rsc_ClusterMon.pid -d -i 30 -E /crm/scripts/crm_mail_agent.sh -h /crm/scripts/crmHtml.html


# change the following values

SENDER="CRMMonitor@customer.com"             ## valid user, else we end in spam or get rejected!
RECEIVER="ralph.roth@customer.com"           ## Additional receiver
HAWKLINK="https://10.66.3.89:7630/cib/live"  ## CHANGEME too!  VIP address

# wait until HTML file is generated successfully
sleep 3

# grep cluster status html file and add a link to Hawklink to it
HTMLTEXT=`cat /crm_scripts/crm_monitor/crmHtml.html`
HAWKLINK="<a href=${HAWKLINK}>Click here to open HAWK Cluster Admin Interface</a><br/><b>Stack:</b>"
HTMLTEXT="${HTMLTEXT/"<b>Stack:</b>"/$HAWKLINK}"

# generate mail subject with causing node, resource + description and date
SUBJ="${CRM_notify_node}: ${CRM_notify_rsc} ${CRM_notify_desc} at $(date) "

# Debug
echo "$(date) - ${CRM_notfiy_node} ${CRM_notify_task}: ${CRM_notify_rsc} ${CRM_notify_desc} ${CRM_notify_rc}" >> /crm_scripts/crm_monitor/crmstate.log

# if [[ unsuccessful monitor operation aka not running ]]
if [[ ${CRM_notify_rc} != 0 && ${CRM_notify_task} == "monitor" && ${CRM_notify_desc} == "not running"  ]]  ## Couldn't parse this if expression. Fix to allow more checks.
then

    # send info to file to reference to it later
    echo "$CRM_notify_node" > /crm_scripts/crm_monitor/crm_current_error
    echo "$CRM_notify_rsc" >> /crm_scripts/crm_monitor/crm_current_error

    # send mail with error
    (
        echo "From: $SENDER "
        echo "To: $RECEIVER "
        echo "MIME-Version: 1.0"
        echo "Content-Type: multipart/alternative; "
        echo ' boundary="ClusterMonitoring by CustomerChangeMe_A2PP"'
        echo "Subject: $SUBJ"
        echo ""
        echo "This is a MIME-encapsulated message"
        echo ""
        echo "--ClusterMonitoring by CustomerChangeMe_A2PP"
        echo "Content-Type: text/html"
        echo ""
        echo "$HTMLTEXT"
    ) | sendmail -t

    # else if [[ successful monitor operation of already failed resource]]
elif [[ ${CRM_notify_task} == "monitor" && ${CRM_notify_desc} == "OK" &&  -f /crm_scripts/crm_monitor/crm_current_error ]]
then

    # read in error reference file
    COUNTER=0
    while IFS='' read -r line || [[ -n "$line" ]]
    do
        if [ "$COUNTER" -eq 0 ]
        then
            LASTNODE=$line
            COUNTER=1
        else
            LASTERROR=$line
        fi
    done < "/crm_scripts/crm_monitor/crm_current_error"

    # check if OK MSG is for reference error
    if [[ "${LASTNODE}" == ${CRM_notify_node} && "${LASTERROR}" == ${CRM_notify_rsc} ]]
    then

        # remove error reference file and clear failed resource
        rm /crm_scripts/crm_monitor/crm_current_error
        crm resource cleanup ${CRM_notify_rsc}

        # send mail with OK MSG
        (
            echo "From: $SENDER "
            echo "To: $RECEIVER "
            echo "MIME-Version: 1.0"
            echo "Content-Type: multipart/alternative; "
            echo ' boundary="ClusterMonitoring by CustomerChangeMe_A2PP"'
            echo "Subject: $SUBJ"
            echo ""
            echo "This is a MIME-encapsulated message"
            echo "--ClusterMonitoring by CustomerChangeMe_A2PP"
            echo "Content-Type: text/html"
            echo ""
            echo "$HTMLTEXT"
        ) | sendmail -t

    fi
    echo "Content-Type: multipart/alternative; "
    echo ' boundary="ClusterMonitoring by ClusterMon"'
    echo "Subject: $SUBJ"
    echo ""
    echo "This is a MIME-encapsulated message"
    echo "--ClusterMonitoring by ClusterMon"
    echo "Content-Type: text/html"
    echo ""
    echo "$HTMLTEXT"
) | sendmail -t

  fi
fi

exit 0
